### PR TITLE
MRG: Fix the computation of the orientation in (..)MxNE and gamma-map

### PR DESCRIPTION
**Pull request by yousrabk**

### CODE REVIEW

File: mne/inverse_sparse/_gamma_map.py [hunk 1]
```diff
@@ -290,21 +292,25 @@ def gamma_map(evoked, forward, noise_cov, alpha, loose="auto", depth=0.8,
...
         if len(X) < 3 * len(active_src):
             X_xyz = np.zeros((3 * len(active_src), X.shape[1]), dtype=X.dtype)
+            X_ori_ = np.zeros((3 * len(active_src), X.shape[1]), dtype=X.dtype)
             for ii in range(len(active_src)):
                 for jj in range(3):
                     if in_pos >= len(active_set):
                         break
                     if (active_set[in_pos] + jj) % 3 == 0:
                         X_xyz[3 * ii + jj] = X[in_pos]
+                        X_ori_[3 * ii + jj] = X_ori[in_pos]
```
Comment (yousrabk) [inline mne/inverse_sparse/_gamma_map.py]: XXX: There is still a problem here, I have to check what is happening when we give `xyz_same_gamma=False`

File: mne/inverse_sparse/_gamma_map.py [hunk 2]
```diff
@@ -277,8 +277,10 @@ def gamma_map(evoked, forward, noise_cov, alpha, loose="auto", depth=0.8,
 
     # Reapply weights to have correct unit
     n_dip_per_pos = 1 if is_fixed_orient(forward) else 3
+    X_ori = X.copy()
     X = _reapply_source_weighting(X, source_weighting,
                                   active_set, n_dip_per_pos)
```
Comment (agramfort) [inline mne/inverse_sparse/_gamma_map.py]: n_dip_per_pos is not used in _reapply_source_weighting :(

File: mne/inverse_sparse/_gamma_map.py [hunk 3]
```diff
@@ -277,8 +277,10 @@ def gamma_map(evoked, forward, noise_cov, alpha, loose="auto", depth=0.8,
...
+    X_ori /= source_weighting[active_set][:, None]
```
Comment (agramfort) [inline mne/inverse_sparse/_gamma_map.py]: this looks bogus. _reapply_source_weighting does `X *= source_weighting[active_set][:, None]``
so I doubt it's doing the right thing but it's just a guess

File: mne/inverse_sparse/_gamma_map.py [hunk 4]
```diff
@@ -290,21 +292,25 @@ def gamma_map(evoked, forward, noise_cov, alpha, loose="auto", depth=0.8,
...
+            X_ori = X_ori_
 
     tmin = evoked.times[0]
     tstep = 1.0 / evoked.info['sfreq']
 
     if return_as_dipoles:
         out = _make_dipoles_sparse(X, active_set, forward, tmin, tstep, M,
-                                   M_estimated, active_is_idx=True)
+                                   M_estimated, active_is_idx=True,
+                                   X_ori=X_ori)
```
Comment (agramfort) [inline mne/inverse_sparse/_gamma_map.py]: it's really a mystery to me why you need to pass both X and X_ori and why internally you use X to get amplitude and X_ori to get orientation.

File: mne/inverse_sparse/tests/test_gamma_map.py [hunk 5]
```diff
@@ -124,5 +124,20 @@ def test_gamma_map_vol_sphere():
 
     assert_array_almost_equal(stc.times, evoked.times, 5)
 
+    # Compare orientation obtained using fit_dipole and mixed_norm
```
Comment (agramfort) [inline mne/inverse_sparse/tests/test_gamma_map.py]: gamma_map

File: mne/inverse_sparse/tests/test_gamma_map.py [hunk 6]
```diff
@@ -124,5 +124,20 @@ def test_gamma_map_vol_sphere():
...
+    # for a simulated evoked
+    pos1 = src[0]['rr'][stc.vertices[0]]
+    pos2 = src[0]['rr'][stc.vertices[1]]
+    evoked_dip = mne.simulation.simulate_evoked(fwd, stc, info, cov, nave=1e9)
+
+    dip_mxne = gamma_map(evoked_dip, fwd, cov, alpha, return_as_dipoles=True)
```
Comment (agramfort) [inline mne/inverse_sparse/tests/test_gamma_map.py]: dip_gmap

File: mne/inverse_sparse/mxne_inverse.py [hunk 7]
```diff
@@ -24,8 +24,7 @@
 @verbose
 def _prepare_weights(forward, gain, source_weighting, weights, weights_min):
     mask = None
-    if isinstance(weights, SourceEstimate):
-        # weights = np.sqrt(np.sum(weights.data ** 2, axis=1))
+    if isinstance(weights, (SourceEstimate, VolSourceEstimate)):
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: if isinstance(weights, _BaseSourceEstimate):

File: mne/inverse_sparse/tests/test_mxne_inverse.py [hunk 8]
```diff
@@ -170,6 +170,23 @@ def test_mxne_vol_sphere():
...
+    pos = src[0]['rr'][stc.vertices[0]]
+    evoked_dip = mne.simulation.simulate_evoked(fwd, stc, info, cov, nave=1e9)
+
+    dip_mxne = mixed_norm(evoked_dip, fwd, cov, alpha=80,
+                          n_mxne_iter=1, maxit=30, tol=1e-8,
+                          active_set_size=10, return_as_dipoles=True)
+
+    amp_max = [np.max(d.amplitude) for d in dip_mxne]
+    dip_mxne = dip_mxne[np.argmax(amp_max)]
+    assert_true(dip_mxne[0].pos[0] in pos)
```
Comment (agramfort) [inline mne/inverse_sparse/tests/test_mxne_inverse.py]: this does not pass on master but the assert below on ori does.

so this test is not enough to demonstrate the bug fix on orientation
estimation.

File: tutorials/plot_brainstorm_phantom_elekta.py [hunk 9]
```diff
@@ -137,3 +138,19 @@
 
 fig.tight_layout()
 plt.show()
+
+
+def plot_pos_ori(pos, ori, color=(0., 0., 0.)):
+    mlab.points3d(pos[0], pos[1], pos[2], scale_factor=0.005,
```
Comment (agramfort) [inline tutorials/plot_brainstorm_phantom_elekta.py]: this works if you pass pos of shape (3, n_times)

as is the example is broken

File: tutorials/plot_brainstorm_phantom_elekta.py [hunk 10]
```diff
@@ -137,3 +138,19 @@
 
 fig.tight_layout()
 plt.show()
+
+
```
Comment (larsoner) [inline tutorials/plot_brainstorm_phantom_elekta.py]: Why not `plot_dipole_locations`? 

https://mne-tools.github.io/dev/generated/mne.viz.plot_dipole_locations.html#mne.viz.plot_dipole_locations

**Comment by agramfort** [comment]
I don't think it works without an MRI...

File: tutorials/plot_brainstorm_phantom_elekta.py (hunk 10 above)
Comment (larsoner) [reply tutorials/plot_brainstorm_phantom_elekta.py]: Ahh okay.

In any case, please make this a separate section with `###########` so it renders nicer in SG

Comment (yousrabk) [reply tutorials/plot_brainstorm_phantom_elekta.py]: Ok

**Comment by yousrabk** [comment]
Are we good here after Eric's comment?

**Comment by larsoner** [comment]
Good enough for me 

https://6100-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.2wP5EEF/html/auto_tutorials/plot_brainstorm_phantom_elekta.html

**Comment by agramfort** [comment]
let me just check tomorrow that the added tests don't work in master so
it's actually obvious as a bug fix

cc @joewalter

**Comment by agramfort** [comment]
ok I did I pass

@yousrabk have a look

FYI @joewalter we realized that the depth weighting strategy for sparse solver was not optimal for orientation estimation and we now use the Frobenius normalization you had written in the past.

File: mne/inverse_sparse/tests/test_gamma_map.py [hunk 11]
```diff
@@ -124,5 +124,22 @@ def test_gamma_map_vol_sphere():
...
+                                                use_cps=True)
+
+    dip_gmap = gamma_map(evoked_dip, fwd, cov, 0.1, return_as_dipoles=True)
+
+    amp_max = [np.max(d.amplitude) for d in dip_gmap]
+    dip_gmap = dip_gmap[np.argmax(amp_max)]
+    assert_true(dip_gmap[0].pos[0] in src[0]['rr'][stc.vertices])
+
+    dip_fit = mne.fit_dipole(evoked_dip, cov, sphere)[0]
+    assert_true(np.abs(np.dot(dip_fit.ori[0], dip_gmap.ori[0])) > 0.99)  # XXX
```
Comment (agramfort) [inline mne/inverse_sparse/tests/test_gamma_map.py]: @yousrabk why the XXX?

Comment (larsoner) [reply mne/inverse_sparse/tests/test_gamma_map.py]: It was in response to my request to add a `# XXX` comment to signify that we should look into whether the `abs` is even necessary at some point (if currents are always positive it shouldn't be), but without at least a few words saying this, it is tough to know that

**Comment by agramfort** [comment]
yes it is necessary since there is no reason why the sign of the amplitude
of the dipoles should be fixed.
please remove

**Comment by larsoner** [comment]
> yes it is necessary since there is no reason why the sign of the amplitude of the dipoles should be fixed.

My point is that we could (for `fit_dipole`) choose to always select the orientation that gives a positive current. Then we wouldn't need it.

**Comment by larsoner** [comment]
(and for `gamma_map` IIUC it uses the source space definition with the dipoles pointing out of cortex, which is the direction of the current, which should be positive)

**Comment by agramfort** [comment]
ok but gamma map can be on a volume / discrete source space so... anyhow if
you keep the XXX please add a comment. I would just remove the XXX

**Comment by larsoner** [comment]
Ahh that's true at least some of these are volume. Sure let's just remove (maybe do it in `master` after merge)

**Comment by agramfort** [comment]
ok XXX removed and what's new updated.

good to go on my end if CIs are happy.

**Comment by larsoner** [comment]
Thanks @yousrabk @agramfort

**Pull request merged**